c7n/resources/ecs.py

```diff
@@ -326,8 +326,76 @@ def process_cluster_resources(self, client, cluster_id, container_instances):
         results = []
         for service_set in chunks(container_instances, self.manager.chunk_size):
             self.manager.log.info(container_instances)
-            results.extend(
-                client.describe_container_instances(
-                    cluster=cluster_id,
-                    containerInstances=container_instances).get('containerInstances', []))
+            r = client.describe_container_instances(cluster=cluster_id,
+                    containerInstances=container_instances).get('containerInstances', [])
+            # Many Container Instance API calls require the cluster_id, adding as a
+            # custodian specific key in the resource
+            for i in r:
+                i['c7n:cluster'] = cluster_id
+            results.extend(r)
         return results
+
+
+@ContainerInstance.action_registry.register('update-state')
+class UpdateState(BaseAction):
+    """Updates a container instance to either ACTIVE or DRAINING
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: drain-container-instances
+              resource: ecs-container-instance
+              actions:
+                - type: update-state
+                  state: DRAINING
+    """
+    schema = type_schema('update-state', state={"type": "string"})
+    permissions = ('ecs:UpdateContainerInstancesState',)
+
+    def process(self, resources):
+        cluster_map = group_by(resources, 'c7n:cluster')
+        for cluster in cluster_map:
+            c_instances = [i['containerInstanceArn'] for i in cluster_map[cluster]]
+            results = self.process_cluster(cluster, c_instances)
+            return results
+
+    def process_cluster(self, cluster, c_instances):
+        # Limit on number of container instance that can be updated in a single
+        # update_container_instances_state call is 10
+        chunk_size = 10
+        client = local_session(self.manager.session_factory).client('ecs')
+        for service_set in chunks(c_instances, chunk_size):
+            try:
+                client.update_container_instances_state(
+                    cluster = cluster,
+                    containerInstances = service_set,
+                    status = self.data.get('state'))
+            except ClientError as e:
```

Comment:
this should catch the no update available error specifically, and continue, perhaps other transients (agent/instance state, unknown instance, etc). for persistent errors we would want this to raise, not swallow the error.

Reply:
it looks like the specific exceptions are `ClientException` `ClusterNotFoundException` `InvalidParameterException` `ServerException`.  They might be sending back the no update status in the `failures` array on a response?  I'm not seeing anything being called out specifically for no update available though.

Reply:
`NoUpdateAvailableException` is found in the exception response: https://github.com/thisisshi/cloud-custodian/blob/aaef49f8375a2c37c540a1fa6189178282680ea5/tests/data/placebo/test_ecs_container_instance_update_unavailable/ecs.UpdateContainerAgent_1.json#L19